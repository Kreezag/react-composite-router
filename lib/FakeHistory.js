export default class FakeHistory {
  constructor () {
    this.location = {
      pathname: '',
      hash: '',
      search: ''
    };

    this.action = '';

    this._listeners = [];
  }

  listen (listener) {
    this._listeners = this._listeners.concat([ listener ]);

    return () => {
      this._listeners = this._listeners.filter((_listener) => _listener !== listener);
    };
  }

  _fire () {
    this._listeners.forEach((listener) => {
      listener(this.location, this.action);
    });
  }

  replace (location) {
    this.location = Object.assign({}, this.location, location);
    this.action = 'REPLACE';

    this._fire();
  }

  push (location) {
    this.location = Object.assign({}, this.location, location);
    this.action = 'PUSH';

    this._fire();
  }

  createHref (location) {
    return `${location.pathname}${location.search ? `?${location.search}` : ''}${location.hash ? `#${location.hash}` : ''}`;
  }
}
