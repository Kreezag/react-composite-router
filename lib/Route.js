import { assertTrimmedNonEmptyString, assertRegExp } from './assert';
import { uniq, each, reduce, isString, isNumber, calcSearchString, parseSearchString } from './utils';


const calcUrlProps = (url) => ({
  path: url.replace(/[?#].*$/, ''),
  searchParams: url.replace(/^[^?]*[?]?/, '').split('&').filter(Boolean)
});


export default class Route {
  constructor (name, { url = '', params = {}, slots = {}, reroute = null } = {}) {
    assertTrimmedNonEmptyString('name', name);
    assertRegExp(/^([a-zA-Z0-9-.]+)$/, 'name', name);

    const parsedUrl = calcUrlProps(url);

    this.url = url;
    this.name = name;
    this.path = parsedUrl.path;
    this.slots = slots;
    this.params = params;
    this.reroute = reroute; // function for replace routes
    this.searchParams = parsedUrl.searchParams;
    this.pathMaskParams = [];

    this.pathMask = this.path.replace(/:([^/]+)/g, ($0, paramName) => {
      this.pathMaskParams.push(paramName);

      return `{{${paramName}}}`;
    });

    this.pathMaskExpSrc = this.path.replace(/:([^/]+)/g, () => '([^/]+)');

    this.pathMaskExp = new RegExp(`^${this.pathMaskExpSrc}$`);

    this.availableUrlParams = this.searchParams.concat(this.pathMaskParams);
    this.additionalParams = Object.keys(this.params)
      .filter((param) => !this.pathMaskParams.includes(param) && !this.searchParams.includes(param));

    this.availableParams = this.availableUrlParams.concat(this.additionalParams);




    // post checking
    if (this.availableUrlParams.length !== uniq(this.availableUrlParams).length) {
      throw new Error(`this "${this.name}" has duplication in params [${this.searchParams.join(',')}] and [${this.pathMaskParams.join(',')}]`);
    }

    const invalidParams = this.additionalParams.filter((param) => !/^[^a-zA-Z]/.test(param));

    if (invalidParams.length) {
      throw new Error(`this "${this.name}" has invalid params [${invalidParams.join(',')}]. additional params must not starts from [a-zA-Z] symbold`);
    }

    const invalidUrlParams = this.availableUrlParams.filter((param) => /^[^a-zA-Z]/.test(param));

    if (invalidUrlParams.length) {
      throw new Error(`this "${this.name}" has invalid url params [${invalidUrlParams.join(',')}]. url params must starts only from [a-z] symbols`);
    }

    each(this.params, (val, key) => {
      if (!this.availableUrlParams.includes(key)) {
        return;
      }

      if (!isString(val) && !isNumber(val)) {
        throw new Error(`this "${this.name}" has invalid param type of "${key}". must be Number or String`);
      }
    });
  }

  isMatch (location, action) {
    return this.pathMaskExp.test(location.pathname || '');
  }

  parseLocationParams (location) { // location
    const { pathname, search, hash } = location;
    let params = Object.assign({}, this.params);

    String(pathname || '').replace(this.pathMaskExp, ($0, ...args) => {
      params = this.pathMaskParams.reduce((params, paramName, index) => {
        params[paramName] = args[index];

        return params;
      }, params);

      return $0;
    });

    const searchParams = search ? parseSearchString(String(search || '').replace(/^\?/, '')) : {};

    params = reduce(searchParams, (params, paramValue, paramName) => {
      if (this.searchParams.includes(paramName)) {
        params[paramName] = paramValue;
      }

      return params;
    }, params);

    return params;
  }

  calcLocation (params) {
    const pathname = this.pathMask.replace(/{{([^}]+)}}/g, ($0, paramName) => {
      if (params.hasOwnProperty(paramName)) {
        return params[paramName];
      }

      throw new Error(`Route param "${paramName}" was not defined for route "${this.name}"`);
    });

    const searchParams = this.searchParams.reduce((search, searchParam) => {
      if (params.hasOwnProperty(searchParam) && params[searchParam] !== this.params[searchParam]) {
        search[searchParam] = params[searchParam];
      }

      return search;
    }, {});

    const location = { pathname, hash: '', search: '' };

    if (searchParams) {
      location.search = calcSearchString(searchParams);
    }

    return location;
  }

  calcParams (stateParams, currentStateParams = {}, resetParams = false) {
    return this.availableParams.reduce((params, paramName) => {
      if (stateParams.hasOwnProperty(paramName)) {
        params[paramName] = stateParams[paramName];
      } else if (!resetParams && currentStateParams.hasOwnProperty(paramName)) {
        params[paramName] = currentStateParams[paramName];
      }

      return params;
    }, Object.assign({}, this.params));
  }

  static createRootRoute (name, definition) {
    const sections = String(name).split('.');

    if (sections.length !== 1) {
      throw new Error(`state "${name}" has invalid format`);
    }

    return new Route(name, definition);
  }

  createChildRoute (name, { url = '', params = {}, slots = {}, reroute = null } = {}) {
    const sections = String(name).split('.');

    if (`${this.name}.${sections[sections.length - 1]}` !== name) {
      throw new Error(`child state name (${name}) must starts from parent name "${this.name}"`);
    }

    const { searchParams, path } = calcUrlProps(url);

    slots = Object.assign({}, this.slots, slots);
    params = Object.assign({}, this.params, params);

    const urlPath = `${this.path.replace(/\/$/, '')}/${path.replace(/^\//, '')}`;
    const urlSearch = `${[ ...this.searchParams, ...searchParams ].join('&')}`;

    url = `${urlPath}${urlSearch ? `?${urlSearch}` : ''}`;

    return new Route(name, { slots, params, url, reroute });
  }
}
