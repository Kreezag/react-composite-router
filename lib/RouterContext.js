import once from 'lodash/once';
import forEach from 'lodash/forEach';
import isEqual from 'lodash/isEqual';
import { assertAvailableProps } from './assert';
import FakeHistory from './FakeHistory';
import EE from './EE';
import Route from './Route';



export default class RouterContext {
  constructor () {
    this.history = new FakeHistory();

    this._listenHistory(this.history);

    this._stateEmitter = new EE();

    this.state = {
      name: '',
      params: {},
      composition: {}
    };

    this.location = { // there is separated location object because history might be in inconsistent state with route
      pathname: '',
      hash: '',
      search: ''
    };

    this.routes = [];
    this.routesByName = {};

    this._hooks = [];
  }

  destroy () {
    this._historyCleanupListener();
    this._stateEmitter.destroy();
    this.location = {};
    this.history = null;
    this.state = {};
    this.routes = [];
    this.routesByName = {};
    this._hooks = [];
  }

  searchStateByLocation (location, action) {
    const route = this.routes.find((route) => route.isMatch(location, action));

    if (!route) {
      return null;
    }

    return {
      name: route.name,
      params: route.parseLocationParams(location)
    };
  }

  setHistory (history) {
    this.history = history;

    this.setLocation(history.location, history.action);

    this._listenHistory(history);
  }

  _listenHistory (history) {
    if (this._historyCleanupListener) {
      this._historyCleanupListener();
    }

    this._historyCleanupListener = once(history.listen((location, action) => {
      this.setLocation(location, action);
    }));
  }

  getLocation () {
    return this.location;
  }

  appendHook (hookFunction) {
    this._hooks = [ ...this._hooks, hookFunction ]; // subscribe

    return once(() => {
      this._hooks = this._hooks.filter((callback) => callback !== hookFunction); // unsubscribe
    });
  }

  applyHooks (nextStateName, nextStateParams) {
    return Promise.all(this._hooks.map((hook) => hook(nextStateName, nextStateParams)));
  }

  setLocation (location, action) {
    const state = this.searchStateByLocation(location, action);

    if (!state) {
      window.console && console.error && console.error(`Router: state is not defined (location: ${JSON.stringify(location)})`); // eslint-disable-line no-console, no-unused-expressions
      return; // not found. rollback to previous state
    }

    if (!this._hooks.length) {
      this.setState(state.name, state.params);
    } else {
      this.applyHooks(state.name, state.params).then(() => {
        this.setState(state.name, state.params);
      });
    }
  }

  setRoutes (routes) {
    forEach(routes, (route) => {
      if (!(route instanceof Route)) {
        throw new TypeError('object is not instance of Route');
      }
    });

    this.routesByName = routes;

    this.routes = Object.keys(routes)
      .sort((name1, name2) => {
        const segmentsLen1 = name1.split('.').length;
        const segmentsLen2 = name2.split('.').length;

        if (segmentsLen1 > segmentsLen2) {
          return -1;
        }

        return segmentsLen1 < segmentsLen2 ? 1 : 0;
      })
      .map((name) => routes[name]);
  }

  setState (name, params = {}) {
    if (this.state.name === name && isEqual(this.state.params, params)) {
      window.console && console.info && console.info('Router: state is the same'); // eslint-disable-line no-console, no-unused-expressions
      return; // state is equal. nothing to do
    }

    if (!this.routesByName.hasOwnProperty(name)) {
      throw new Error(`invalid state name "${name}"`);
    }

    this.state = {
      name,
      params,
      composition: this.routesByName[name].slots
    };

    this.location = this.calcLocation(name, params);

    this._stateEmitter.publish(name, params);
  }

  onStateChange (listenerFunc) {
    return this._stateEmitter.subscribe(listenerFunc);
  }

  moveTo (location, { reload = false, replace = false } = {} /*{ replace, reload }*/) {
    if (!location) {
      throw new Error('location is empty');
    }

    if (!reload) {
      if (replace) {
        this.history.replace(location);
      } else {
        this.history.push(location);
      }

      return;
    }

    window.location.href = this.history.createHref(location);
  }

  createHref (location) {
    return this.history.createHref(location);
  }

  calcLocation (stateName, stateParams, { reset = false } = {}) {
    if (!this.routesByName.hasOwnProperty(stateName)) {
      throw new Error(`RouterContext: state "${stateName}" not found`);
    }

    const route = this.routesByName[stateName];
    const params = route.calcParams(stateParams, this.state.params, reset);

    if (process.env.NODE_ENV !== 'production') { // eslint-disable-line no-process-env
      assertAvailableProps(route.availableParams, 'route params', params);
    }

    if (route.reroute) {
      const newState = route.reroute({ name: route.name, params }, { name: this.state.name, params: this.state.params });

      if (newState !== undefined) {
        if (this.routesByName.hasOwnProperty(newState.name)) {
          const newRoute = this.routesByName[newState.name];

          return newRoute.calcLocation(Object.assign({}, newRoute.params, newState.params || {}));
        }

        throw new Error('invalid route after rerouting');
      }
    }

    return route.calcLocation(params);
  }

  getState () {
    return this.state;
  }
}
