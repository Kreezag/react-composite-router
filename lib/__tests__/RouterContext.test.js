/* eslint-env browser, jest */

import RouterContext from '../RouterContext';
import Route from '../Route';
import FakeHistory from '../FakeHistory';
import History from './__mocks__/history';

describe('RouterContext with FakeHistory', () => {
  it('should call onChange', () => {
    const context = new RouterContext();

    context.setRoutes({
      'some.state': new Route('some.state', { url: '/some-state?some', params: { _hello: 3 } }),
      'some.state.other': new Route('some.state.other', { url: '/some-other?some', params: { _hello: 3 } }),
      'some': new Route('some', { url: '/' })
    });

    const history = new FakeHistory();

    history.location.pathname = '/';

    context.setHistory(history);

    const listener = jest.fn();

    const destroy = context.onStateChange(listener);

    history.replace({ pathname: '/some-state' });

    expect(listener).toBeCalled();

    context.destroy();

    destroy();
  });

  it('should call moveTo', () => {
    const context = new RouterContext();

    context.setRoutes({
      'some.state': new Route('some.state', { url: '/some-state?some', params: { _hello: 3 } }),
      'some.state.other': new Route('some.state.other', { url: '/some-other?some', params: { _hello: 3 } }),
      'some': new Route('some', { url: '/' })
    });

    const listener = jest.fn();

    context.onStateChange(listener);

    context.moveTo({ pathname: '/', hash: '', search: '' });
    context.moveTo({ pathname: '/some-other', hash: '', search: '' }, { replace: true });

    expect(listener).toBeCalled();
  });

  it('should calc location with rerouting', () => {
    const context = new RouterContext();

    context.setRoutes({
      'some.state': new Route('some.state', { url: '/some-state?some', params: { _hello: 3 } }),
      'some.state.other': new Route('some.state.other', { url: '/some-other?some', params: { _hello: 3 } }),
      'some': new Route('some', { url: '/', reroute: () => ({ name: 'some.state.other' }) })
    });

    expect(context.calcLocation('some', {})).toMatchSnapshot();

    expect(context.calcLocation('some.state', {})).toMatchSnapshot();

    expect(context.calcLocation('some.state.other', {})).toMatchSnapshot();
  });

  it('should calc location', () => {
    const context = new RouterContext();

    context.setRoutes({
      root: new Route('root', { url: '/' }),
      'some-state': new Route('some-state', { url: '/some-state?some', params: { _hello: 3 } })
    });

    context.setState('some-state', { some: 4 });

    const location = context.calcLocation('some-state', { some: 3 });

    const href = context.createHref(location);

    expect(location).toMatchSnapshot();
    expect(context).toMatchSnapshot();
    expect(href).toBe('/some-state?some=3');
  });
});

describe('RouterContext with new history', () => {

  it('', () => {
    const context = new RouterContext();

    context.setRoutes({
      'some-state': new Route('some-state', { url: '/some-state?some', params: { _hello: 3 } })
    });

    const result1 = context.searchStateByLocation({ pathname: '', hash: '', search: '' }, 'POP');

    expect(result1).toMatchSnapshot();

    const result2 = context.searchStateByLocation({ pathname: '/some-state', hash: '', search: '' }, 'POP');

    expect(result2).toMatchSnapshot();
  });

  it('should run every listener', () => {
    const context = new RouterContext();

    context.setRoutes({
      'some-state': new Route('some-state', { url: '/some-state?some', params: { _hello: 3 } })
    });

    const func1 = jest.fn();
    const func2 = jest.fn();
    const func3 = jest.fn();

    context.onStateChange(func1);
    context.onStateChange(func2);
    const off = jest.fn(context.onStateChange(func3));

    off();

    context.setState('some-state');

    expect(func1).toBeCalled();
    expect(func2).toBeCalled();
    expect(func3).not.toBeCalled();
    expect(off).toBeCalled();
    expect(context).toMatchSnapshot();
  });

  it('should calc location', () => {
    const context = new RouterContext();

    context.setRoutes({
      root: new Route('root', { url: '/' }),
      'some-state': new Route('some-state', { url: '/some-state?some', params: { _hello: 3 } })
    });

    context.setHistory(History({ pathname: '/' }));

    context.setState('some-state', { some: 4 });

    const location = context.calcLocation('some-state', { some: 3 });

    const href = context.createHref(location);

    expect(location).toMatchSnapshot();
    expect(context).toMatchSnapshot();
    expect(href).toBe('/some-state?some=3');
  });

  it('should destroy', () => {
    const context = new RouterContext();

    const func1 = jest.fn();
    const func2 = jest.fn();
    const func3 = jest.fn();

    context.onStateChange(func1);

    context.destroy();

    context.onStateChange(func2);

    context.setRoutes({
      'some-state': new Route('some-state', { url: '/some-state?some', params: { _hello: 3 } })
    });

    context.setState('some-state');

    context.destroy();

    context.onStateChange(func3);

    context.setRoutes({
      'some-state': new Route('some-state', { url: '/some-state?some', params: { _hello: 3 } })
    });

    context.setState('some-state');

    expect(func1).not.toBeCalled();
    expect(func2).toBeCalled();
    expect(func3).toBeCalled();
    expect(context).toMatchSnapshot();
  });
});
