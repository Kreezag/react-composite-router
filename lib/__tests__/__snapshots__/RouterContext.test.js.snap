exports[`RouterContext with FakeHistory should calc location 1`] = `
Object {
  "hash": "",
  "pathname": "/some-state",
  "search": "some=3",
}
`;

exports[`RouterContext with FakeHistory should calc location 2`] = `
RouterContext {
  "_changeListeners": Array [],
  "_historyCleanupListener": [Function],
  "history": FakeHistory {
    "_listeners": Array [
      [Function],
    ],
    "action": "",
    "location": Object {
      "hash": "",
      "pathname": "",
      "search": "",
    },
  },
  "location": Object {
    "hash": "",
    "pathname": "/some-state",
    "search": "some=4",
  },
  "routes": Array [
    Route {
      "additionalParams": Array [],
      "availableParams": Array [],
      "availableUrlParams": Array [],
      "name": "root",
      "params": Object {},
      "path": "/",
      "pathMask": "/",
      "pathMaskExp": /^\\/$/,
      "pathMaskExpSrc": "/",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [],
      "slots": Object {},
      "url": "/",
    },
    Route {
      "additionalParams": Array [
        "_hello",
      ],
      "availableParams": Array [
        "some",
        "_hello",
      ],
      "availableUrlParams": Array [
        "some",
      ],
      "name": "some-state",
      "params": Object {
        "_hello": 3,
      },
      "path": "/some-state",
      "pathMask": "/some-state",
      "pathMaskExp": /^\\/some-state$/,
      "pathMaskExpSrc": "/some-state",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [
        "some",
      ],
      "slots": Object {},
      "url": "/some-state?some",
    },
  ],
  "routesByName": Object {
    "root": Route {
      "additionalParams": Array [],
      "availableParams": Array [],
      "availableUrlParams": Array [],
      "name": "root",
      "params": Object {},
      "path": "/",
      "pathMask": "/",
      "pathMaskExp": /^\\/$/,
      "pathMaskExpSrc": "/",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [],
      "slots": Object {},
      "url": "/",
    },
    "some-state": Route {
      "additionalParams": Array [
        "_hello",
      ],
      "availableParams": Array [
        "some",
        "_hello",
      ],
      "availableUrlParams": Array [
        "some",
      ],
      "name": "some-state",
      "params": Object {
        "_hello": 3,
      },
      "path": "/some-state",
      "pathMask": "/some-state",
      "pathMaskExp": /^\\/some-state$/,
      "pathMaskExpSrc": "/some-state",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [
        "some",
      ],
      "slots": Object {},
      "url": "/some-state?some",
    },
  },
  "state": Object {
    "composition": Object {},
    "name": "some-state",
    "params": Object {
      "some": 4,
    },
  },
}
`;

exports[`RouterContext with FakeHistory should calc location with rerouting 1`] = `
Object {
  "hash": "",
  "pathname": "/some-other",
  "search": "",
}
`;

exports[`RouterContext with FakeHistory should calc location with rerouting 2`] = `
Object {
  "hash": "",
  "pathname": "/some-state",
  "search": "",
}
`;

exports[`RouterContext with FakeHistory should calc location with rerouting 3`] = `
Object {
  "hash": "",
  "pathname": "/some-other",
  "search": "",
}
`;

exports[`RouterContext with new history  1`] = `null`;

exports[`RouterContext with new history  2`] = `
Object {
  "name": "some-state",
  "params": Object {
    "_hello": 3,
  },
}
`;

exports[`RouterContext with new history should calc location 1`] = `
Object {
  "hash": "",
  "pathname": "/some-state",
  "search": "some=3",
}
`;

exports[`RouterContext with new history should calc location 2`] = `
RouterContext {
  "_changeListeners": Array [],
  "_historyCleanupListener": [Function],
  "history": Object {
    "createHref": [Function],
    "listen": [Function],
    "location": Object {
      "hash": "",
      "pathname": "/",
      "search": "",
    },
    "push": [Function],
    "replace": [Function],
  },
  "location": Object {
    "hash": "",
    "pathname": "/some-state",
    "search": "some=4",
  },
  "routes": Array [
    Route {
      "additionalParams": Array [],
      "availableParams": Array [],
      "availableUrlParams": Array [],
      "name": "root",
      "params": Object {},
      "path": "/",
      "pathMask": "/",
      "pathMaskExp": /^\\/$/,
      "pathMaskExpSrc": "/",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [],
      "slots": Object {},
      "url": "/",
    },
    Route {
      "additionalParams": Array [
        "_hello",
      ],
      "availableParams": Array [
        "some",
        "_hello",
      ],
      "availableUrlParams": Array [
        "some",
      ],
      "name": "some-state",
      "params": Object {
        "_hello": 3,
      },
      "path": "/some-state",
      "pathMask": "/some-state",
      "pathMaskExp": /^\\/some-state$/,
      "pathMaskExpSrc": "/some-state",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [
        "some",
      ],
      "slots": Object {},
      "url": "/some-state?some",
    },
  ],
  "routesByName": Object {
    "root": Route {
      "additionalParams": Array [],
      "availableParams": Array [],
      "availableUrlParams": Array [],
      "name": "root",
      "params": Object {},
      "path": "/",
      "pathMask": "/",
      "pathMaskExp": /^\\/$/,
      "pathMaskExpSrc": "/",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [],
      "slots": Object {},
      "url": "/",
    },
    "some-state": Route {
      "additionalParams": Array [
        "_hello",
      ],
      "availableParams": Array [
        "some",
        "_hello",
      ],
      "availableUrlParams": Array [
        "some",
      ],
      "name": "some-state",
      "params": Object {
        "_hello": 3,
      },
      "path": "/some-state",
      "pathMask": "/some-state",
      "pathMaskExp": /^\\/some-state$/,
      "pathMaskExpSrc": "/some-state",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [
        "some",
      ],
      "slots": Object {},
      "url": "/some-state?some",
    },
  },
  "state": Object {
    "composition": Object {},
    "name": "some-state",
    "params": Object {
      "some": 4,
    },
  },
}
`;

exports[`RouterContext with new history should destroy 1`] = `
RouterContext {
  "_changeListeners": Array [
    [Function],
  ],
  "_historyCleanupListener": [Function],
  "history": FakeHistory {
    "_listeners": Array [
      [Function],
    ],
    "action": "",
    "location": Object {
      "hash": "",
      "pathname": "",
      "search": "",
    },
  },
  "location": Object {
    "hash": "",
    "pathname": "/some-state",
    "search": "",
  },
  "routes": Array [
    Route {
      "additionalParams": Array [
        "_hello",
      ],
      "availableParams": Array [
        "some",
        "_hello",
      ],
      "availableUrlParams": Array [
        "some",
      ],
      "name": "some-state",
      "params": Object {
        "_hello": 3,
      },
      "path": "/some-state",
      "pathMask": "/some-state",
      "pathMaskExp": /^\\/some-state$/,
      "pathMaskExpSrc": "/some-state",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [
        "some",
      ],
      "slots": Object {},
      "url": "/some-state?some",
    },
  ],
  "routesByName": Object {
    "some-state": Route {
      "additionalParams": Array [
        "_hello",
      ],
      "availableParams": Array [
        "some",
        "_hello",
      ],
      "availableUrlParams": Array [
        "some",
      ],
      "name": "some-state",
      "params": Object {
        "_hello": 3,
      },
      "path": "/some-state",
      "pathMask": "/some-state",
      "pathMaskExp": /^\\/some-state$/,
      "pathMaskExpSrc": "/some-state",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [
        "some",
      ],
      "slots": Object {},
      "url": "/some-state?some",
    },
  },
  "state": Object {
    "composition": Object {},
    "name": "some-state",
    "params": Object {},
  },
}
`;

exports[`RouterContext with new history should run every listener 1`] = `
RouterContext {
  "_changeListeners": Array [
    [Function],
    [Function],
  ],
  "_historyCleanupListener": [Function],
  "history": FakeHistory {
    "_listeners": Array [
      [Function],
    ],
    "action": "",
    "location": Object {
      "hash": "",
      "pathname": "",
      "search": "",
    },
  },
  "location": Object {
    "hash": "",
    "pathname": "/some-state",
    "search": "",
  },
  "routes": Array [
    Route {
      "additionalParams": Array [
        "_hello",
      ],
      "availableParams": Array [
        "some",
        "_hello",
      ],
      "availableUrlParams": Array [
        "some",
      ],
      "name": "some-state",
      "params": Object {
        "_hello": 3,
      },
      "path": "/some-state",
      "pathMask": "/some-state",
      "pathMaskExp": /^\\/some-state$/,
      "pathMaskExpSrc": "/some-state",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [
        "some",
      ],
      "slots": Object {},
      "url": "/some-state?some",
    },
  ],
  "routesByName": Object {
    "some-state": Route {
      "additionalParams": Array [
        "_hello",
      ],
      "availableParams": Array [
        "some",
        "_hello",
      ],
      "availableUrlParams": Array [
        "some",
      ],
      "name": "some-state",
      "params": Object {
        "_hello": 3,
      },
      "path": "/some-state",
      "pathMask": "/some-state",
      "pathMaskExp": /^\\/some-state$/,
      "pathMaskExpSrc": "/some-state",
      "pathMaskParams": Array [],
      "reroute": null,
      "searchParams": Array [
        "some",
      ],
      "slots": Object {},
      "url": "/some-state?some",
    },
  },
  "state": Object {
    "composition": Object {},
    "name": "some-state",
    "params": Object {},
  },
}
`;
